src/irmin-pack/irmin_pack.ml

```diff
@@ -557,8 +577,76 @@ struct
         in
         Contents.CA.sync ~on_generation_change (contents_t t)
 
-      (** Storea share instances so one clear is enough. *)
+      (** Stores share instances so one clear is enough. *)
       let clear t = Contents.CA.clear (contents_t t)
+
+      (** Migrate data from the IO [src] (with [name] in path [root_old]) into
+          the temporary dir [root_tmp], then swap in the replaced version. *)
+      let migrate_io_to_v2 ~root_old ~root_tmp ~progress (name, src) =
+        let old, tmp = (root_old // name, root_tmp // name) in
+        IO.upgrade ~progress ~src ~dst:(tmp, `V2) |> function
+        | Ok () ->
+            Platform.rename tmp old;
```

Comment:
I am wondering if we could make that API a bit more simple by doing the rename in `IO.upgrade`. That way we wouldn't expose `tmp` at all (e.g. you could just derive a temporary filename from `src`)

Reply:
Yep; it's simpler this way – thanks.

IOs now derive their own `tmp` file locations individually, and rename internally.